--- a/asio/include/asio/detail/thread_info_base.hpp
+++ b/asio/include/asio/detail/thread_info_base.hpp
@@ -142,6 +142,8 @@
 
     if (this_thread)
     {
+      int best_index = -1;
+      std::size_t best_chunks = UCHAR_MAX + std::size_t(1);
       for (int mem_index = Purpose::begin_mem_index;
           mem_index < Purpose::end_mem_index; ++mem_index)
       {
@@ -150,13 +152,23 @@
           void* const pointer = this_thread->reusable_memory_[mem_index];
           unsigned char* const mem = static_cast<unsigned char*>(pointer);
           if (static_cast<std::size_t>(mem[0]) >= chunks
+              && static_cast<std::size_t>(mem[0]) < best_chunks
               && reinterpret_cast<std::size_t>(pointer) % align == 0)
           {
-            this_thread->reusable_memory_[mem_index] = 0;
-            mem[size] = mem[0];
-            return pointer;
+            best_index = mem_index;
+            best_chunks = mem[0];
+            if (best_chunks == chunks)
+              break;
           }
         }
+      }
+      if (best_index >= 0)
+      {
+        void* const pointer = this_thread->reusable_memory_[best_index];
+        unsigned char* const mem = static_cast<unsigned char*>(pointer);
+        this_thread->reusable_memory_[best_index] = 0;
+        mem[size] = mem[0];
+        return pointer;
       }
 
       for (int mem_index = Purpose::begin_mem_index;
